### PR TITLE
Table: Add select/unselect all column values to table filter

### LINKED ISSUE #79371: [FeatureRequest] Table column filter by values "check all" functionality
### Discussed in https://github.com/grafana/grafana/discussions/38279

<div type='discussions-op-text'>

<sup>Originally posted by **memfiz** May 13, 2021</sup>
**What would you like to be added**:

Column filter by values needs select all functionality. Similarly, it has a "Clear filter" when you want to clear all checked values.

**Why is this needed**:

When using big tables there is often a need to filter by values to select many entries. Then you use the filter you can find e.g. 100 of values and there is no possibility to select them all with one click, you need to check them one by one, that is not feasible with such a large number of entries.

As an example
[https://play.grafana.org/d/baqo21CGz/column-filter-select-all-request?orgId=1](https://play.grafana.org/d/baqo21CGz/column-filter-select-all-request?orgId=1)
If you would like to select values of _id column that contain "5ndM" there is no way to check them all with one click after the filter finds them.

</div>

### PR DESCRIPTION
**Pull request by ahuarte47**
**What is this feature?**

This MR adds two buttons to easily select or unselect subset of columns values in Table panels:

+ New Select values button: 
   To select the subset of column values currently displayed.
+ New Unselect values button:
   To unselect the subset of column values currently displayed.

**Why do we need this feature?**

When using big tables there is often a need to filter by values to select many entries. Then you use the filter you can find e.g. 100 of values and there is no possibility to select them all with one click, you need to check them one by one, that is not feasible with such a large number of entries.

**Who is this feature for?**

Users filtetring large list of values in tables.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/79371

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: docs/sources/panels-visualizations/visualizations/table/index.md [hunk 1]
```diff
@@ -191,6 +191,18 @@ To filter column values, click the filter (funnel) icon next to a column title.
 
 Click the check box next to the values that you want to display. Enter text in the search field at the top to show those values in the display so that you can select them rather than scroll to find them.
 
+### Select column values
```
Comment (imatwawana) [inline docs/sources/panels-visualizations/visualizations/table/index.md:194]: If I'm understanding correctly, this is a function to select or deselect all the visible values and add them to your filtering. 
```suggestion
### Select all displayed column values
```

File: docs/sources/panels-visualizations/visualizations/table/index.md [hunk 2]
```diff
@@ -191,6 +191,18 @@ To filter column values, click the filter (funnel) icon next to a column title.
...
+
+To select the subset of column values currently displayed.
+
+Click the **Select values** button to add the current subset of visible values to the full set of filter column values.
```
Comment (imatwawana) [inline docs/sources/panels-visualizations/visualizations/table/index.md:198]: I'm trying to simplify the language a bit here. Hopefully I haven't changed the meaning.
```suggestion
Select all displayed column values.

Click the **Select values** button to add all displayed values to the filter.
```

File: docs/sources/panels-visualizations/visualizations/table/index.md [hunk 3]
```diff
@@ -191,6 +191,18 @@ To filter column values, click the filter (funnel) icon next to a column title.
...
+
+### Unselect column values
```
Comment (imatwawana) [inline docs/sources/panels-visualizations/visualizations/table/index.md:200]:
```suggestion
### Clear selected column values 
```

File: docs/sources/panels-visualizations/visualizations/table/index.md [hunk 4]
```diff
@@ -191,6 +191,18 @@ To filter column values, click the filter (funnel) icon next to a column title.
...
+
+To unselect the subset of column values currently displayed.
+
+Click the **Unselect values** button to remove the current subset of visible values from the full set of filter column values.
```
Comment (imatwawana) [inline docs/sources/panels-visualizations/visualizations/table/index.md:204]:
```suggestion
Remove the all displayed column values from the filter selection.

Click the **Unselect values** button to remove all displayed column values from the filter.
```

**Review by imatwawana — COMMENTED** [review]
Thanks for the contribution! I'm just reviewing the documentation updates, so there will still need to be a technical review. I've made a couple grammar edits but mostly tried to simplify some of the language. I think we probably also need to think about changing the names of the buttons, but I'd need to see the implementation first (hopefully when all the checks pass, the system will provide a preview).

File: docs/sources/panels-visualizations/visualizations/table/index.md (hunk 1 above)
Comment (ahuarte47) [reply docs/sources/panels-visualizations/visualizations/table/index.md:194]: Yes, this is the idea, thanks you very much for your changes!

**Comment by ahuarte47** [comment]
thank you very much @imatwawana for you review and help.

**Review by oscarkilhed — COMMENTED** [review]
Hey, thank you for your contribution!

image

This is useful and the code looks good at first glance (I'll look closer soon). But I think we could improve the UI and the wording on these buttons. @catherineymgui, is this something to bring to one of the UX feedback sessions?

**Comment by catherineymgui** [comment]
Yes, please bring it to a UX feedback session!

**Comment by ahuarte47** [comment]
Hi there, please, may I propose a new feature in this PR? I have added a new ListBox to alllow users to select what items are displayed using comparison expressions, not only using regex patterns.

It is useful when you need filter values like numbers.

This new commit show a list box (In order to do not alter the behavior in other panels, there is in the component a new `showOperators` flag disabled by default).

Screenshot from 2023-12-13 18-44-12

Screenshot from 2023-12-13 18-44-46

I still need to fix an issue with the `Select` component, I do not know why it closes its parent `FilterPopup` component when you click in the list of operations. Please, do you know why?

What do you think? It can be useful to include? If you disagree, I revert this change.
Thanks in advance

**Comment by tjanson** [comment]
@ahuarte47 I'm merely a humble user, but I've been looking for this kind of feature for ages. It seems clearly outside the scope of this issue, but while we're at it:

* For numeric values, it would be extremely useful to filter by ranges (`score > 7`, `3 < score < 7`),
* for text values a simple search would be useful (`description contains "openssl"`),
* for timestamps, time ranges (`date before 2022-02-02`), and so on.

This can be extended further and further (why not auto-bucket numeric values? arbitrary regular expressions?), but what you propose would already be hugely beneficial.

**Comment by ahuarte47** [comment]
> @ahuarte47 I'm merely a humble user, but I've been looking for this kind of feature for ages. It seems clearly outside the scope of this issue, but while we're at it:
> 
>     * For numeric values, it would be extremely useful to filter by ranges (`score > 7`, `3 < score < 7`),
> 
>     * for text values a simple search would be useful (`description contains "openssl"`),
> 
>     * for timestamps, time ranges (`date before 2022-02-02`), and so on.
> 
> 
> This can be extended further and further (why not auto-bucket numeric values? arbitrary regular expressions?), but what you propose would already be hugely beneficial.

Hi @tjanson this extra filter could be implemented in new operator "expression", but I am not an expert grafana & react developer, any help would be great! I hope I haven't already made a mess in the code.

**Comment by oscarkilhed** [comment]
I think we should keep this PR to the original scope. Smaller incremental improvements are generally preferred.

I brought this change to the UX feedback session and the general consensus was that a UX similar to the github inbox would work better.

image
image

**Comment by ahuarte47** [comment]
> I think we should keep this PR to the original scope. Smaller incremental improvements are generally preferred.
> 
OK, great! I remove that extra temporal commit.

**Comment by reindlt** [comment]
Hi, seems to be a similiar functionality as in #79072. If this one is accepted I think the other one could be omitted.

**Comment by ahuarte47** [comment]
> I brought this change to the UX feedback session and the general consensus was that a UX similar to the github inbox would work better.

Hi, I have refactorized this PR, now it is working with a three-state Checkbox:
image

**Comment by oscarkilhed** [comment]
@ahuarte47  Thank you, I will get back to this soon!

**Comment by imatwawana** [comment]
@ahuarte47 - After @oscarkilhed has completed his review and any changes are implemented, please request review from me again, so I can take another look at the docs with the updated UI. Just FYI, I'll be away until January 2nd after today, but I'll add a label so the PR doesn't go stale during the holidays.

**Comment by ahuarte47** [comment]
> @ahuarte47 I'm merely a humble user, but I've been looking for this kind of feature for ages. It seems clearly outside the scope of this issue, but while we're at it:
> 
>     * For numeric values, it would be extremely useful to filter by ranges (`score > 7`, `3 < score < 7`),
> 
>     * for text values a simple search would be useful (`description contains "openssl"`),
> 
>     * for timestamps, time ranges (`date before 2022-02-02`), and so on.
> 
> 
> This can be extended further and further (why not auto-bucket numeric values? arbitrary regular expressions?), but what you propose would already be hugely beneficial.

I have created a new feature-request issue: https://github.com/grafana/grafana/issues/79800 I will try to propose a MR soon.

**Review by oscarkilhed — COMMENTED** [review]
I think this is close to mergable. 

I've created a pull request (https://github.com/ahuarte47/grafana/pull/1) to your branch to align the select all checkbox with the rest of the checkboxes.

I also removed the "Click to" bit in the checkbox label as I found it superfluous.

**Review by oscarkilhed — APPROVED** [review]
This looks good. When @imatwawana has gone over the docs we can merge this.

File: docs/sources/panels-visualizations/visualizations/table/index.md [hunk 5]
```diff
@@ -191,6 +191,8 @@ To filter column values, click the filter (funnel) icon next to a column title.
 
 Click the check box next to the values that you want to display. Enter text in the search field at the top to show those values in the display so that you can select them rather than scroll to find them.
 
+Click the check box next to the Ok/Cancel buttons to add or remove all displayed values to/from the filter.
```
Comment (imatwawana) [inline docs/sources/panels-visualizations/visualizations/table/index.md:194]: In the screenshot, it looks like the box is positioned above buttons.
```suggestion
Click the check box above the **Ok** and **Cancel** buttons to add or remove all displayed values to/from the filter.
```

**Review by imatwawana — APPROVED** [review]
One quibble about where the box is located. If I'm wrong, please update my suggestion accordingly and commit to add the text formatting for the buttons.

**Comment by ahuarte47** [comment]
Thank you very much @imatwawana  @oscarkilhed for your help!

**Comment by ahuarte47** [comment]
Thanks @oscarkilhed @imatwawana for Your help!

**Comment by imatwawana** [comment]
Thanks for the contribution, @ahuarte47! I was literally wishing for this feature just a couple weeks ago!

**Comment by Andrew713** [comment]
Hi! The feature looks amazing and it is just what I was looking for. However, I have trouble implementing the same changes to my local Grafana instance. 

I simply deployed `image: grafana/grafana-enterprise` via docker-compose, and my assumption was that I can simply edit the 'packages/grafana-ui/src/components/Table/FilterList.tsx' file directly inside of my grafana container. But **I can't find it anywhere in container's file system**, so I might be doing something wrong (aside from the whole approach of manually editing code inside running container)

Version is Grafana v10.2.3 (1e84fede54)

Do you have any ideas?

**Comment by ahuarte47** [comment]
Hi @Andrew713 in my understanding Grafana codebase is compiled to build the final app using `yarn`. `tsx` files are not deployed but the compiled version of them.

**Comment by Andrew713** [comment]
> Hi @Andrew713 in my understanding Grafana codebase is compiled to build the final app using `yarn`. `tsx` files are not deployed but the compiled version of them.
> 
> https://github.com/grafana/grafana/blob/main/contribute/developer-guide.md#build-grafana

@ahuarte47 You are absolutely right, I had the wrong idea of this. Probably should just rebuild fresh with this feature

**Pull request merged**